text/0436-gamelift-l2.md

```diff
@@ -0,0 +1,999 @@
+
+* The file name and path of an executable in your game build.
+
+* Optionally Parameters to pass to the process on launch.
+
+* The number of processes to run concurrently.
+
+A GameLift instance is limited to 50 processes running concurrently.
+
+```ts fixture=with-build
+import * as gamelift from '@aws-cdk-lib/aws-gamelift';
+
+const fleet = new gamelift.Fleet(this, 'Game server fleet', {
+  build: build,
+  runtimeConfiguration: {
+    gameSessionActivationTimeoutSeconds: 123,
+    maxConcurrentGameSessionActivations: 123,
+  }
+});
+fleet.runtimeConfiguration.addServerProcess({
+  launchPath: '/local/game/GameLiftExampleServer.x86_64', 
+  parameters: '-logFile /local/game/logs/myserver1935.log -port 1935',
+  concurrentExecutions: 100,
+});
+```
+
+See [Managing how game servers are launched for hosting](https://docs.aws.amazon.com/gamelift/latest/developerguide/fleets-multiprocess.html)
+in the *Amazon GameLift Developer Guide*.
+
+##### Defining an instance type
+
+GameLift uses Amazon Elastic Compute Cloud (Amazon EC2) resources, called instances, to deploy your game servers and host game sessions for your players. When setting up a new fleet, you decide what type of instances your game needs and how to run game server processes on them (using a runtime configuration). All instances in a fleet use the same type of resources and the same runtime configuration. You can edit a fleet's runtime configuration and other fleet properties, but the type of resources cannot be changed.
+
+```ts fixture=with-build
+import * as gamelift from '@aws-cdk-lib/aws-gamelift';
+import * as ec2 from '@aws-cdk-lib/aws-ec2';
+
+new gamelift.Fleet(this, 'Game server fleet', {
+  build: build,
+  instanceType: ec2.InstanceType.of(ec2.InstanceClass.C5, ec2.InstanceSize.LARGE)
+});
+```
+
+##### Using Spot instances
+
+When setting up your hosting resources, you have the option of using Spot Instances, On-Demand Instances, or a combination.
+
+By default, this property is set to ON_DEMAND.
+
+```ts fixture=with-build
+import * as gamelift from '@aws-cdk-lib/aws-gamelift';
+
+new gamelift.Fleet(this, 'Game server fleet', {
+  build: build,
+  type: FleetType.SPOT
+});
+```
+
+##### Allowing Ingress traffic
+
+The allowed IP address ranges and port settings that allow inbound traffic to access game sessions on this fleet.
+
+New game sessions are assigned an IP address/port number combination, which must fall into the fleet's allowed ranges. Fleets with custom game builds must have permissions explicitly set. For Realtime Servers fleets, GameLift automatically opens two port ranges, one for TCP messaging and one for UDP.
```

Comment:
Ingress port is not binded to a game code (build or script) but to a runtimeConfiguration as we can override default game server port through a dedicated server process configuration.

Reply:
Can you elaborate? When would you need to specify an ingress _without_ using a build or vice versa.

Reply:
a game build is a game server which can be run multiple times depending on the runtimeConfiguration. 

For example, this configuration:

```
const fleet = new gamelift.Fleet(this, 'Game server fleet', {
  content: build,
  gameSessionActivationTimeoutSeconds: 123,
  maxConcurrentGameSessionActivations: 123,
  serverProcesses: [{
    launchPath: '/local/game/GameLiftExampleServer.x86_64', 
    parameters: '-logFile /local/game/logs/myserver1935.log -port 1935',
  }, {
    launchPath: '/local/game/GameLiftExampleServer.x86_64', 
    parameters: '-logFile /local/game/logs/myserver1936.log -port 1936',
  }, {
    launchPath: '/local/game/GameLiftExampleServer.x86_64', 
    parameters: '-logFile /local/game/logs/myserver1937.log -port 1937',
  }],
  ingress: [
      {
        peer: ec2.Peer.anyIpv4(),
        fromPort: ec2.Port.tcp(1935),
        toPort: ec2.Port.tcp(1937),
      },
    ],
});
```

Reply:
ok I see. I think what we need to do is make it hard for users to mess this up. I think what we don't want is a bunch of properties that are only required in certain scenarios.

In this case I might forget to provide `ingress` with a build fleet or I might provide ingress with a script fleet. Or in the example above I need to remember to provide the port in two different locations.

I'm not sure what the best solution is, maybe we need to have a `FleetBase` and then a `BuildFleet` and a `ScriptFleet`?

@otaviomacedo curious what you think here.

Reply:
I was thinking the same way at first but the service team disagree with that proposal as it do not fit with their Api and Console experience.

Having ingress configured is mandatory for a GameServerBuild and automatically settled when using a realtimescript.

Reply:
Any update here?

Reply:
> Having ingress configured is mandatory for a GameServerBuild and automatically settled when using a realtimescript.

Then I agree with Cory. We should have different subclasses so that users can only provide an ingress when it's mandatory (`BuildFleet`).

Reply:
Agreed, that was also my first choice. Will make all the change and create 2 subclasses `ScriptFleet` and `BuildFleet`